app/Models/BeatmapPack.php

```diff
@@ -74,65 +76,56 @@ public function userCompletionData($user)
-                                    }
-                                });
-                        });
-                    }
-                });
-            } else {
-                $modeStr = Beatmap::modeStr($this->playmode);
-
-                if ($modeStr === null) {
-                    throw new Exception("beatmapset pack {$this->getKey()} has invalid playmode: {$this->playmode}");
-                }
-
-                $scoreRelation = camel_case("scores_best_{$modeStr}");
-
-                $query->whereHas($scoreRelation, function ($query) use ($userId) {
-                    $query->where('user_id', '=', $userId);
-
-                    if ($this->no_diff_reduction) {
-                        $query->withoutMods(app('mods')->difficultyReductionIds->toArray());
-                    }
-                });
+
+            $beatmaps = Beatmap
+                ::whereIn('beatmapset_id', $this->items()->select('beatmapset_id'))
+                ->select(['beatmap_id', 'beatmapset_id', 'playmode'])
+                ->get();
+            $beatmapsetIdsByBeatmapId = [];
+            foreach ($beatmaps as $beatmap) {
+                $beatmapsetIdsByBeatmapId[$beatmap->beatmap_id] = $beatmap->beatmapset_id;
+            }
+            $params = [
+                'beatmap_ids' => array_keys($beatmapsetIdsByBeatmapId),
+                'exclude_converts' => $this->playmode === null,
+                'limit' => 0,
+                'ruleset_id' => $this->playmode,
+                'user_id' => $userId,
+            ];
+            if ($this->no_diff_reduction) {
+                $params['exclude_mods'] = app('mods')->difficultyReductionIds->toArray();
             }
 
-            $completedBeatmapsetIds = $query->pluck('beatmapset_id')->all();
-            $completed = count($completedBeatmapsetIds) === count($beatmapsetIds);
+            static $aggName = 'by_beatmap';
+            static $subAggScore = 'scores';
+
+            $search = new ScoreSearch(ScoreSearchParams::fromArray($params));
+            $search->size(0);
+            $search->setAggregations([$aggName => [
+                'terms' => [
+                    'field' => 'beatmap_id',
+                    'size' => max(1, count($params['beatmap_ids'])),
+                ],
+                'aggs' => [
+                    $subAggScore => [
+                        'top_hits' => [
+                            'size' => 1,
+                        ],
+                    ],
+                ],
+            ]]);
+            $response = $search->response();
+            $search->assertNoError();
+            $scoreIds = array_map(
+                fn (array $hit): string => $hit[$subAggScore]['hits']['hits'][0]['_id'],
+                $response->aggregations($aggName)['buckets'],
+            );
+            $completedBeatmapIds = Solo\Score::whereKey($scoreIds)->pluck('beatmap_id')->all();
```

Comment:
Can't this use the `key` values from `$response->aggregations($aggName)['buckets']` instead?